taar_lite/recommenders/cache.py

```diff
@@ -0,0 +1,114 @@
+import redis
+import boto3
+import json
+from srgutil.interfaces import IMozLogging
+
+
+class LazyJSONLoader:
+    def __init__(self, ctx, s3_bucket, s3_key, expiry=14400):
+        self._ctx = ctx
+
+        if not self.ignore_redis():
+            self._redis = redis.StrictRedis(host='localhost', port=6379, db=0)
+
+        self.logger = self._ctx[IMozLogging].get_logger('srgutil')
+
+        self._s3_bucket = s3_bucket
+        self._s3_key = s3_key
+
+        self._expiry_seconds = expiry
+        self._cached_copy = None
+
+    def ignore_redis(self):
+        return self._ctx.get('ignore_redis', False)
+
+    def force_expiry(self):
+        if self.ignore_redis():
+            return
+
+        key_str = "{}|{}".format(self._s3_bucket, self._s3_key)
+
+        # Clear the byte cache in redis as well as locally cached copy
+        self._redis.delete(key_str)
+        self._cached_copy = None
+
+    def get(self):
+        """Fetch the value that is stored"""
+
+        if self.ignore_redis():
+            return self._get_json()
+
+        key_str = "{}|{}".format(self._s3_bucket, self._s3_key)
+
+        # If we have no cached copy, we explicitly want to clear the
+        # byte cache in redis
+        if self._cached_copy is None:
```

Comment:
I'm not sure there is - in the first case - the cached_copy is cold - usually because this is first startup so we've never copied the redis cached copy into local memory.  The second case is that the redis cached copy has expired (the key no longer exists in redis) in which case we need to clobber the locally cached copy as well as the redis cached copy from raw S3.

It's probably not entirely clear, but the redis cache is used for two things:
* the taarlite server should not cause unnecessary hits to the S3 server if we're under load
* the redis cache should expire in a timely manner and the locally cached copy should be clobbered with whatever is in S3 currently

Reply:
TTRYWR = 45min

Reply:
*sigh*. victor is wrong.  just ignore me.

Reply:
![image](https://user-images.githubusercontent.com/4829/40391726-473510b2-5de7-11e8-9fa4-48f890959029.png)